pmemcheck/pmc_tx.c

```diff
@@ -0,0 +1,458 @@
+}
+
+/**
+ * \brief Remove the given thread entry from the active thread register.
+ * \param[in,out] thread_entry The thread entry to be removed.
+ */
+static void
+remove_thread_entry(struct thread_info *thread_entry)
+{
+    /* sanity check */
+    if (!VG_(OSetGen_Contains)(trans.threads, thread_entry))
+        return;
+
+    /* remove and destroy entry */
+    VG_(OSetGen_Remove)(trans.threads, thread_entry);
+    VG_(OSetWord_Destroy)(thread_entry->tx_ids);
+    VG_(OSetGen_FreeNode)(trans.threads, thread_entry);
+}
+
+/**
+ * \brief Remove a transaction
+ * \param[in] tx_id Transaction id to remove - must be previously registered.
+ */
+UInt
+remove_tx(UWord tx_id)
+{
+    struct tx_info *tx = VG_(OSetGen_Lookup)(trans.transactions, &tx_id);
+
+    if (tx == NULL)
+        return 1;
+
+    --(tx->counter);
+
+    if (tx->counter > 0)
+        return 0;
+
+    /* remove the transaction from each thread */
+    struct thread_info *elem;
+    VG_(OSetGen_ResetIter)(trans.threads);
+    while ((elem = VG_(OSetGen_Next)(trans.threads)) != NULL) {
+        VG_(OSetWord_Remove)(elem->tx_ids, tx_id);
+        /* remove thread entry if it does not have any more active txs */
+        if (VG_(OSetWord_Size)(elem->tx_ids) == 0)
+            remove_thread_entry(elem);
+    }
+
+    VG_(OSetGen_Destroy)(tx->regions);
+    VG_(OSetGen_Remove)(trans.transactions, &tx_id);
+    VG_(OSetGen_FreeNode)(trans.transactions, tx);
+
+    return 0;
+}
+
+/**
+ * \brief Check if the running thread contributes to a given transaction.
+ * \param[in] tx_id The id of the transaction.
+ * \return True if thread contributes to the transaction, false otherwise.
+ */
+static Bool
+is_tx_in_thread(UWord tx_id)
+{
+   /* search all transactions this thread is in */
+    UWord thread_id = VG_(get_running_tid)();
+    struct thread_info *tinfo= VG_(OSetGen_Lookup)(trans.threads, &thread_id);
+    /* thread not part of any transaction */
+    if (tinfo == NULL) {
+        if (trans.verbose)
+            VG_(dmsg)("thread %lu not part of any transaction\n", thread_id);
+        return False;
+    }
+
+    return VG_(OSetWord_Contains)(tinfo->tx_ids, tx_id);
+}
+
+/**
+ * \brief Add a memory region to a transaction.
+ * \param[in] tx_id The id of the transaction the object will be added to.
+ * \param[in] base The starting address of the region to be added.
+ * \param[in] size The size of the added region.
+ * \return -1 when tx_id is invalid, -2 when the thread does not contribute
+ *            to the given transaction, 0 otherwise.
+ */
+UInt
+add_obj_to_tx(UWord tx_id, UWord base, UWord size)
+{
+    struct tx_info *tx = VG_(OSetGen_Lookup)(trans.transactions, &tx_id);
+    if (tx == NULL) {
+        /* no matching transaction found */
+        if (trans.verbose)
+            VG_(dmsg)("no matching transaction found\n");
+        return 1;
+    }
+
+
+
+    /* this thread does not participate in this transaction */
+    if (!is_tx_in_thread(tx_id)) {
```

Comment:
So, even with explicit transactions, *pmemcheck* would treat it as illegal if two threads contribute to the same TX (assuming they're are properly synchronized)?